cranelift-module/src/dwarf.rs

```diff
@@ -0,0 +1,295 @@
+use cranelift_codegen::isa::{RegInfo, RegUnit, TargetIsa};
+use cranelift_codegen::{ir, isa};
+
+use std::boxed::Box;
+use std::string::String;
+use std::string::ToString;
+use std::vec::Vec;
+
+use gimli;
+use gimli::constants::{DW_EH_PE_pcrel, DW_EH_PE_sdata4};
+use gimli::write::Address;
+use gimli::write::CallFrameInstruction;
+use gimli::write::CommonInformationEntry;
+use gimli::write::EhFrame;
+use gimli::write::Error;
+use gimli::write::FrameDescriptionEntry;
+
+const PC_SDATA4: u8 = DW_EH_PE_pcrel.0 | DW_EH_PE_sdata4.0;
+
+/// FrameSink maintains state for and an interface to construct DWARF frame information for a
+/// cranelift-produced module.
+pub struct FrameSink {
+    // we need to retain function names to hand out usize identifiers for FDE addresses,
+    // which are then used to look up function names again for relocations, when `write_address` is
+    // called.
+    fn_names: Vec<String>,
```

Comment:
`ModuleContents` does already contain something like this, not sure if it is worth trying to reuse that, but maybe `FuncId::as_u32` can be the identifier.